### PR TITLE
Improve accessibility

### LINKED ISSUE #137: screen reader support
from https://community.kobotoolbox.org/t/accessibility-standards-for-visually-disabilities/10885

When filling out the form using a screen reader programme:

* The voice-over reads content that is not actually visible on the page. I assume the programme reads something from the code of the page, including content from other pages. The voice-over should read only the content that is visible on the page.
* When moving to the next page, screen reader by default goes to the bottom of a page. It is very inconvenient when a user with visual impairment must manually search and find the beginning of a page. When clicking “next”, it should take you to the next question by default.

### LINKED ISSUE #735: screen reader support
from https://community.kobotoolbox.org/t/accessibility-standards-for-visually-disabilities/10885

When filling out the form using a screen reader programme:

* The voice-over reads content that is not actually visible on the page. I assume the programme reads something from the code of the page, including content from other pages. The voice-over should read only the content that is visible on the page.
* When moving to the next page, screen reader by default goes to the bottom of a page. It is very inconvenient when a user with visual impairment must manually search and find the beginning of a page. When clicking “next”, it should take you to the next question by default.

### PR DESCRIPTION
**Pull request by FrankApiyo**
closes enketo/enketo#137

**NB:** These changes were tested manually on chrome using the *[chromevox](https://chrome.google.com/webstore/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn?hl=en)* screen reader/extension

### CODE REVIEW

**Review by MartijnR — CHANGES_REQUESTED** [review]
Thanks @FrankApiyo!

File: src/js/page.js [hunk 1]
```diff
@@ -367,7 +367,17 @@ export default {
...
+            pageEl.focus();
         }
+
+        document.querySelectorAll( '[tabindex="0"]' ).forEach( el => el.setAttribute( 'tabIndex', -1 ) );
+        const nextPageButton = document.querySelector( '.next-page' );
+        if ( nextPageButton ) {
+            nextPageButton.setAttribute( 'tabIndex', 2 );
+        }
+        document.querySelectorAll( 'input[readonly]' ).forEach( el => el.setAttribute( 'aria-hidden', 'true' ) );
+        document.querySelectorAll( '[style*="display: none"]' ).forEach( el => el.setAttribute( 'aria-hidden', 'true' ) );
```
[the remark this answers is not in the document]
Comment (MartijnR) [reply src/js/page.js:380]: Actually, should readonly questions be hidden at all? E.g. a readonly question with a default value or calculated value?

Perhaps you'd like to only hide so-called empty readonly questions without calculations or dynamic defaults? Then definitely look at readonly.js to add this there.

File: src/js/page.js [hunk 2]
```diff
@@ -367,7 +367,17 @@ export default {
             this._focusOnFirstQuestion( pageEl );
             this._toggleButtons( newIndex );
             pageEl.dispatchEvent( events.PageFlip() );
+            pageEl.setAttribute( 'tabIndex', 1 );
+            pageEl.focus();
```
Comment (MartijnR) [inline src/js/page.js:371]: Did you look into why `_focusOnFirstQuestion` is not working?

**Review by MartijnR — CHANGES_REQUESTED** [review]

File: src/js/page.js [hunk 3]
```diff
@@ -367,7 +367,17 @@ export default {
             this._focusOnFirstQuestion( pageEl );
             this._toggleButtons( newIndex );
             pageEl.dispatchEvent( events.PageFlip() );
+            pageEl.setAttribute( 'tabIndex', 1 );
```
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/js/page.js:370]: oversight! fixing this

File: src/js/page.js (hunk 2 above)
Comment (FrankApiyo) [reply src/js/page.js:371]: It seems that `_focusOnFirstQuestion` calls `trigger( 'fakefocus' );` .  After looking around for a while, it seems `fakefocus` is not handled anywhere.

File: src/js/page.js [hunk 4]
```diff
@@ -367,7 +367,17 @@ export default {
...
+            pageEl.focus();
         }
+
+        document.querySelectorAll( '[tabindex="0"]' ).forEach( el => el.setAttribute( 'tabIndex', -1 ) );
+        const nextPageButton = document.querySelector( '.next-page' );
+        if ( nextPageButton ) {
+            nextPageButton.setAttribute( 'tabIndex', 2 );
+        }
```
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/js/page.js:378]: going to move it to `init()`

File: src/js/readonly.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ export default {
         const nodes = this.form.getRelatedNodes( 'readonly', '', updated ).get();
         nodes.forEach( node =>  {
             node.closest( '.question' ).classList.add( 'readonly' );
+            node.setAttribute( 'aria-hidden', 'true' );
```
Comment (MartijnR) [inline src/js/readonly.js:15]: Just repeating my earlier question. Should all readonly inputs really be aria-hidden? It's very common to show a calculated value in a readonly input field. I just want to make sure.

File: src/js/page.js [hunk 6]
```diff
@@ -29,6 +29,7 @@ export default {
      * @type {Function}
      */
     init() {
+        document.querySelectorAll( '[tabindex="0"]' ).forEach( el => el.setAttribute( 'tabIndex', -1 ) );
```
Comment (MartijnR) [inline src/js/page.js:32]: I don't feed good about this kind of DOM correction. Makes more sense to fix at the source, or am I overlooking something?

File: src/js/page.js [hunk 7]
```diff
@@ -65,6 +66,11 @@ export default {
                 this._setLangChangeHandlers();
                 this.active = true;
                 this._flipToFirst();
+
+                const nextPageButton = document.querySelector( '.next-page' );
```
Comment (MartijnR) [inline src/js/page.js:70]: see $btnNext a few lines above

File: src/js/page.js (hunk 2 above)
Comment (MartijnR) [reply src/js/page.js:377]: Well, it seems to me that focusOnFirstQuestion's purpose should be to focus on the first question... :)  So it makes sense to me to fix that instead. 

page.js should actually be de-jquerified (eventually). Listening should be done with `events.FakeFocus().type` and triggering with `events.FakeFocus()`. It looks like it's used to update the progress bar only.

Surprisingly this function only appears to be scrolling to the first question and not focusing at all. I don't remember why, but let's add the focus there! 

Note that a full-page form also doesn't focus on the first question. Is that also a problem with screen readers?

**Review by MartijnR — CHANGES_REQUESTED** [review]

**Comment by lognaturel** [comment]
> Note that a full-page form also doesn't focus on the first question. Is that also a problem with screen readers?

Forgive me, I'm jumping in here without a lot of context but this happened to catch my eye. I've found this slightly annoying myself. On a computer pressing tab once gets the expected focus but that's not very discoverable. On mobile it would be very nice to have the keyboard out and ready to go when starting a new record.

A related issue that I find even more problematic is what happens after submission in a non-single context. We're left at the bottom of the screen and have to manually scroll up and put focus on the first question to start a new record. Forms can have hundreds or thousands of questions so this can add a lot of friction. Even knowing this behavior I get confused so I expect it's an accessibility issue. I would expect that after a successful submission, the page scrolls to the top and the first question has focus. If this is worth following up on but doesn't feel related to this work, I'm happy to file a separate issue.

**Comment by MartijnR** [comment]
thanks @lognaturel!

> e left at the bottom of the screen and have to manually scroll up and put focus on the first question 

Interestingly, I never realized that and am not sure if this ever worked! I've added an issue here just in case, for scrolling: https://github.com/enketo/enketo-express/issues/248 (the scrolling likely needs a code change in enketo-express).

Wrt to focusing on the first question, I hadn't thought of the keyboard popping up on mobile browsers. I would have thought that undesirable (and that's maybe why we don't have this). So good to know that seems like a helpful thing to you, and I'd have no problem introducing this. We can always easily remove it again if it causes issues.

So if you agree @FrankApiyo, and if it is within your scope, please focus a one-page form on the first question as well.

File: src/js/readonly.js (hunk 5 above)
Comment (FrankApiyo) [reply src/js/readonly.js:15]: I'm not certain if all read-only fields are hidden, but it seems those with *class* `empty` are hidden for sure https://github.com/enketo/enketo-core/blob/78a971943c2e1b4d7c85c618a6522c8be538941e/src/sass/core/_main.scss#L297

going to change this to check if *class* `empty` is present in the node

File: src/js/page.js (hunk 6 above)
Comment (FrankApiyo) [reply src/js/page.js:32]: it seems this line isn't needed after all; after testing these changes again with a screen reader, it seems focus stays were it's expected to without this line

File: src/js/page.js (hunk 7 above)
Comment (FrankApiyo) [reply src/js/page.js:70]: `$btnNext` seemed to be missing the `setAttribute` method

Comment (FrankApiyo) [reply src/js/page.js:70]: going to use `attr` method

File: src/js/page.js (hunk 2 above)
Comment (FrankApiyo) [reply src/js/page.js:377]: moving the call to `.focus()` to the *'fakefocus'* event handler makes sense. however, I'm not sure why the event handler is not called at all in spite of the even being fired. this is happening for me locally.

**Review by MartijnR — CHANGES_REQUESTED** [review]

File: src/js/readonly.js [hunk 8]
```diff
@@ -20,6 +20,10 @@ export default {
             const empty = !node.value && !node.dataset.calculate && !setValue && !node.classList.contains( 'readonly-forced' );
 
             node.classList.toggle( 'empty', empty );
+
+            if( node.classList.contains( 'empty' ) ){
+                node.setAttribute( 'aria-hidden', 'true' );
+            }
```
Comment (MartijnR) [inline src/js/readonly.js:26]: Doesn't this mean the attribute will remain there forever (even if it becomes non-empty)?

Also `if(empty)` would probably be faster.

File: src/js/form.js [hunk 9]
```diff
@@ -790,7 +790,11 @@ Form.prototype.setEventHandlers = function() {
         // update the form progress status
         this.progress.update( event.target );
     } );
+
     this.view.html.addEventListener( events.FakeFocus().type, event => {
+        // focus on element
+        event.target.focus();
+
```
[the remark this answers is not in the document]
Comment (MartijnR) [reply src/js/form.js:797]: Can this be removed now, that you've moved it somewhere else?

**Review by MartijnR — CHANGES_REQUESTED** [review]

File: src/js/readonly.js (hunk 8 above)
Comment (FrankApiyo) [reply src/js/readonly.js:28]: going to remove `aria-hidden`  attribute when not `empty` is false

**Review by MartijnR — APPROVED** [review]

**Review by MartijnR — APPROVED** [review]
Thanks!

**Pull request merged**